src/OrchardCore/OrchardCore.Search.Elasticsearch.Core/Services/ElasticIndexInitializerService.cs

```diff
@@ -0,0 +1,67 @@
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+using OrchardCore.Environment.Shell;
+using OrchardCore.Environment.Shell.Models;
+using OrchardCore.Environment.Shell.Scope;
+using OrchardCore.Modules;
+using OrchardCore.Search.Elasticsearch.Core.Services;
+
+namespace OrchardCore.Search.Elasticsearch
+{
+    /// <summary>
+    /// Provides a way to initialize Elasticsearch index on startup of the application
+    /// if the index is not found on the Elasticsearch server.
+    /// </summary>
+    public class ElasticIndexInitializerService : IModularTenantEvents
+    {
+        private readonly ShellSettings _shellSettings;
+
+        public ElasticIndexInitializerService(ShellSettings shellSettings)
+        {
+            _shellSettings = shellSettings;
+        }
+
+        public Task ActivatedAsync()
+        {
+            if (_shellSettings.State == TenantState.Running)
+            {
+                ShellScope.AddDeferredTask(async scope =>
+                {
+                    var elasticIndexSettingsService = scope.ServiceProvider.GetRequiredService<ElasticIndexSettingsService>();
+                    var elasticIndexingService = scope.ServiceProvider.GetRequiredService<ElasticIndexingService>();
+                    var indexManager = scope.ServiceProvider.GetRequiredService<ElasticIndexManager>();
+
+                    var elasticIndexSettings = await elasticIndexSettingsService.GetSettingsAsync();
+
+                    foreach (var settings in elasticIndexSettings)
+                    {
+                        if (!await indexManager.Exists(settings.IndexName))
+                        {
+                            await elasticIndexingService.CreateIndexAsync(settings);
+
+                            // TODO: fix execution of shellscope
+                            //await elasticIndexingService.ProcessContentItemsAsync(settings.IndexName);
```

Comment:
@jtkech for some reason this fails to execute. I'm not sure why but the ElasticIndexingService is registered as a Scoped service so normally it should work.

Reply:
So here the shell itself is not yet marked as activated, yes it is executed in a deferred task but that is related to the shellScope we are using to activate the shell, so if you use another `shellScope.UsingAsync()` you should specify `activateShell: false` as it is done for Lucene, otherwise it will re-run the activate events in an infinite recursion.

Reply:
Done, thanks 😄